theano/gof/fg.py

```diff
@@ -320,13 +321,15 @@ def __remove_clients__(self, r, clients_to_remove,
                               if output.clients or output in self.outputs]
             # If the apply node is not used and is not an output
             if not used_or_output:
-                self.apply_nodes.remove(apply_node)
```

Comment:
@nouiz, sorry, I had missed out this comment while addressing the rest. Would make this along with the change that you've suggested now.